### PR TITLE
Improve config final validation

### PR DESCRIPTION
**Pull request by OttoWinter**
# What does this implement/fix? 

Refactor the "post-validation" functionality in #1631 a bit.

The function is definitely very useful for the esphome codebase, and enables better validation at compile time.

This PR mainly changes the code behind the functionality to:

1. convert it to use voluptuous - this has a number of advantages, for example all `esphome.config_validation` validators can be used. But this also enables us to show better context, as now the validator can use the `path` argument to say where the error is.
2. Rename `validate` to `POST_VALIDATE_SCHEMA` - I think `validate` is too broad, isn't `CONFIG_SCHEMA` also `validate` etc. Not 100% happy with the new name either, but it expresses that this is a hook that happens _after_ (post) the main validation.
3. Refactored `declare_ids` variable to not be local, for other developers it's hard to see the local variable actually mutates the `config` state too.
4. Create new `types.py` and `post_validate.py` modules. Former to prevent cyclic imports, latter for an abstract base class for better typing.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
> a bit

🤣 this is a good one

> expresses that this is a hook that happens after (post) the main validation

but this is still main validation, is last step. Name is wrong because we can still validate


----

May be it more makes sense to just add this to `CONFIG_SCHEMA`

The validators for uart, e.g. coud look like this:

```py
    .extend(cv.polling_component_schema("60s"))
    .extend(uart.UART_DEVICE_SCHEMA( "cse7766", baud_rate=4800, require_rx=True))
)
```

then this somehow forwards the parameters...
```py
# A schema to use for all UART devices, all UART integrations must extend this!
UART_DEVICE_SCHEMA = cv.Schema(
    {
        cv.GenerateID(CONF_UART_ID): cv.all(cv.use_id(UARTComponent), pv.id_declaration_match_schema(validate_hub with the args)
    }
)
```

The idea is if it is possible to create a validator that will run later, with access to full config, so the validator argument can really be a v function that will run in the last stage, then you won't need `POST_VALIDATE_SCHEMA`

also a generic last validator can be added, e.g. rtttl?

**Comment by OttoWinter** [comment]
> :rofl: this is a good one
yeah sorry it grew a lot bigger than I wanted :P

> but this is still main validation, is last step. Name is wrong because we can still validate

I see your point, yeah. But on the other hand `validate` makes it sound like you can use it to validate just like with `CONFIG_SCHEMA` (where the two are quite different).

> May be it more makes sense to just add this to CONFIG_SCHEMA

Hmm, good idea! Probably won't be easy to add tbh though, there'd need to be some way to track the registered ~~post~~ whatever-validation somewhere. I'll think about it to see if there's a way.

(tracking in the schema itself probably wouldn't work well, because if it's wrapped in `cv.All`/other validators we'd need to recover it somewhere; and in the validated return value too, if any validator afterwards mutates the value it gets forgotten)

> also a generic last validator can be added, e.g. rtttl?

Oops, I forgot that component, thanks! Please elaborate `generic last validator`, not sure what you mean by that

**Comment by glmnet** [comment]
> I see your point, yeah. But on the other hand validate makes it sound like you can use it to validate just like with CONFIG_SCHEMA (where the two are quite different).

Agree `validate` is to broad, just wanted to point `POST` might be confusing, maybe something with "final"  as in a try .. finally like `FINAL_VALIDATE_SCHEMA`

> Hmm, good idea! Probably won't be easy to add tbh though, there'd need to be some way to track the registered post whatever-validation somewhere. I'll think about it to see if there's a way.

I'd give it a try, no time now and will take me ages :) but looks funny

> Oops, I forgot that component, thanks! Please elaborate generic last validator, not sure what you mean by that

I mean just a validator that not depends on a key like CONF_UART_ID, or CONF_OUTPUT in rtttl, just a generic validator for use in e.g. the wifi validation that Jesse added earlier

**Comment by OttoWinter** [comment]
> maybe something with "final" as in a try .. finally like FINAL_VALIDATE_SCHEMA

that's a great name, thanks!

File: esphome/components/rtttl/__init__.py [hunk 1]
```diff
@@ -55,14 +52,25 @@ def validate(config, item_config):
     ]
 
     if platform in PWM_BAD:
-        raise ValueError(f"Component rtttl cannot use {platform} as output component")
+        raise cv.Invalid(f"Component rtttl cannot use {platform} as output component")
 
     if platform not in PWM_GOOD:
         _LOGGER.warning(
-            "Component rtttl is not known to work with the selected output type. Make sure this output supports custom frequency output method."
+            "Component rtttl is not known to work with the selecteValueErrort type. "
```
Comment (oxan) [inline esphome/components/rtttl/__init__.py:59]: Looks like something was accidentally dragged in here.

**Comment by jesserockz** [comment]
Do we want to change the validate I added in `wifi` too in this PR?

https://github.com/esphome/esphome/blob/607c3ae65151dd791a02c92d92eac08ffff2baaa/esphome/components/wifi/__init__.py#L140-L149

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Thanks @OttoWinter

**Pull request merged**